threads/src/main/java/org/jboss/as/threads/ThreadFactoryResolver.java

```diff
@@ -140,8 +140,12 @@ private ServiceName resolveDefaultThreadFactory(String threadPoolName, ServiceNa
                                                         ServiceTarget serviceTarget) {
             final ServiceName threadFactoryServiceName = threadPoolServiceName.append("thread-factory");
             final ThreadFactoryService service = new ThreadFactoryService();
-            service.setThreadGroupName(getThreadGroupName(threadPoolName));
-            service.setNamePattern("%G - %t");
+
+            // Do not set service#threadGroupName, make the group name part of the pattern instead. Setting the
+            // service#threadGroupName would lead to a ThreadGroup being created each time the service is restarted.
+            String threadGroupName = getThreadGroupName(threadPoolName);
+            service.setNamePattern(threadGroupName + " - %t");
+
```

Comment:
I think this is the most problematic part - related to the ThreadFactoryResourceDefinition / ThreadFactoryService.

This resource allows to create custom thread factory, and allows user to define:
* thread group name,
* thread priority
* thread name pattern (which can use %G to emit thread group name).

The thread priority is not in fact set through a created ThreadGroup instance but directly on newly created Threads, so the ThreadGroup still doesn't have much meaning.

Above is a workaround - do not set the threadGroupName on the ThreadFactoryService, so no ThreadGroup is going to be created, instead use the thread group name as part of the name pattern.

If we have a freedom change the resource model, we could remove the threadGroupName attribute from the ThreadFactoryResource and just leave the name pattern and priority attributes, which still allows for the same end effect (thread group name can be "hard-coded" part of the pattern).

Reply:
@TomasHofman What you've done here is fine, except please add @Deprecated(forRemoval="true") to AbstractThreadFactoryResolver.getThreadGroupName

Also, I think ThreadFactoryService should cache any ThreadGroup created in start in an instance field and call ThreadGroup.destroy() on it in stop().  Catch and ignore or DEBUG log any RuntimeException thrown from destroy(). If whatever service injected the ThreadFactoryService didn't clean up in its own stop() the threads it created, the destroy() call will fail, but oh well. Lets not fail or make log noise but we can debug log. If they did clean up destroy() will clean up the parent TG ref to the group.

The various 'xxxThreadPoolService' classes in this module clean up their threads correctly.

Reply:
@bstansberry @TomasHofman `threadGroup.destroy()` is deprecated for removal on JDK 17 so I discourage their usage. I agree with moving it to a  variable instance since there is a 1:1 relationship between the service and the ThreadGroup, at first glance, I am expecting that a service reload can reuse the same ThreadGroup.

I commented on the JBoss List as well, the removal of the ThreadsGroups, although it looks like the way to follow since this class has been degraded in JDK, seems that it could be helpful for debugging purposes, so I am more inclined to keep them on the services that require it as a static field instead of remove them altogether. At least, don't do it until there is any other available option on the JDK to group them.

Thoughts?

Reply:
Hi @yersan, I did see the ticket you referred on the mailing list (https://bugs.openjdk.org/browse/JDK-8252885), it mentions some debugging related APIs like JDWP, JVM TI and JDI which I think is what you meant. It's sort of like these APIs contains some methods that lets you enumerate threads and subgroups belonging to a ThreadGroup.

My perspective is that we would not loose anything that much important. If you can't filter by groups you can still filter by name. The tools that I use don't support this anyway, or I didn't notice because I never felt the need to use it... But I know this is very subjective statement.

Regarding keeping an instance field in the `ThreadFactoryService` for created group, I think that only makes sense if we want to call the destroy method on it. The `ThreadFactoryService` instances are created fresh after each app server reload, so it won't help with reusing the original groups. If we want to reuse the original groups we would have to keep them in some some static map with service name or something like that as a key, or we would need to keep the `ThreadFactoryService` instance itself in a static field wherever they are created (and then have the instance field for the group).

The destroy method is still present in the JDK 21, I'm now checking, but it is an empty method. I'm slightly in favor of not using it, but I will go by what you guys like most.

Reply:
I'll defer to @yersan re the static ThreadGroups.

I don't see anything useful IDEA does with them; it just displays the TG name as part of its string representation of a thread. I don't see any filtering or sorting feature.  I pinged @ehsavoie to ask if NetBeans does anything.

JConsole also doesn't do anything (i.e. the platform mbeans don't show TG data.)

If destroy() is a no-op in SE 21, then +1 to not using it. It being deprecated in 17 is good enough reason to drop it. And I agree if we don't call destroy() using an instance field does nothing. So for that we either live with the leak or add a static WeakHashMap<String, ThreadGroup> to cache them.

Reply:
NetBeans shows threads grouped under their thread group in the debug window

Reply:
In eclipse IDE you can group the threads by ThreadGroup in your debug view, so hence the utility for keeping the ThreadGroups if possible but removing the leak.